src/log.js

```diff
@@ -1,153 +1,389 @@
-      return flatten(res).concat(diff)
-    })
-  }
-
-  _insert(entry) {
-    let indices = entry.next.map((next) => this._items.map((f) => f.hash).indexOf(next)) // Find the item's parent's indices
-    const index = indices.length > 0 ? Math.max(Math.max.apply(null, indices) + 1, 0) : 0 // find the largest index (latest parent)
-    this._items.splice(index, 0, entry)
-    return entry
+  /**
+   * Get the log in JSON format
+   * @returns {Object<{heads}>}
+   */
+  toJSON() {
+    return { heads: this.heads }
   }
 
-  _commit() {
-    this._items = this._items.concat(this._currentBatch)
-    this._currentBatch = []
+  /**
+   * Get the log as a Buffer
+   * @returns {Buffer}
+   */
+  toBuffer() {
+    return new Buffer(JSON.stringify(this.toJSON()))
   }
+}
 
-  _fetchRecursive(ipfs, hash, all, amount, depth) {
-    const isReferenced = (list, item) => list.reverse().find((f) => f === item) !== undefined
-    let result = []
+class LogUtils {
+  /**
+   * Create a new log
+   * @param {IPFS} ipfs An IPFS instance
+   * @param {string} id - ID for the log
+   * @param {Array} [entries] - Entries for this log
+   * @param {Array} [heads] - Heads for this log
+   * @returns {Log}
+   */
+  static create(ipfs, entries, heads) {
+    if (!ipfs) throw new Error('Ipfs instance not defined')
+      // TODO: if (Array.isArray(entries))
+      // TODO: entries.map((e) => Entry.isEntry(e) ? e : await Entry.create(ipfs, e))
+    return new Log(ipfs, entries, heads)
+  }
 
-    this.events.emit('progress', 1)
+  /**
+   * Create a new log starting from an entry
+   * @param {IPFS} [ipfs] An IPFS instance
+   * @param {string} id
+   * @param {string} [hash] [Multihash of the entry to start from]
+   * @param {Number} length
+   * @param {function(hash, entry, parent, depth)} onProgressCallback
+   * @returns {Promise<Log>}
+   */
+  static fromEntry(ipfs, hash, length = -1, onProgressCallback) {
+    return LogUtils._fetchRecursive(ipfs, hash, {}, length, 0, null, onProgressCallback)
+      .then((items) => {
+        let log = LogUtils.create(ipfs)
+        items.reverse().forEach((e) => LogUtils._insert(ipfs, log, e))
+        log._heads = LogUtils._findHeads(log)
+        return log
+      })
+  }
 
-    // If the given hash is in the given log (all) or if we're at maximum depth, return
-    if (isReferenced(all, hash) || depth >= amount)
-      return Promise.resolve(result)
+  /**
+   * Create a log from multihash
+   * @param {IPFS} [ipfs] An IPFS instance
+   * @param {string} [hash] Multihash to create the log from
+   * @param {Number} [length=-1] How many items to include in the log
+   * @returns {Promise<Log>}
+   */
+  static fromMultihash(ipfs, hash, length = -1, onProgressCallback) {
+    if (!ipfs) throw new Error('Ipfs instance not defined')
+    if (!hash) throw new Error('Invalid hash: ' + hash)
+    let logData
+    return ipfs.object.get(hash, { enc: 'base58' })
+      .then((res) => logData = JSON.parse(res.toJSON().data))
```

Comment:
`logData` is only used in the next promise below, you could just return it rather than pull it out of the scope